### PR TITLE
Deletion of Model Versions 

### PR DESCRIPTION
**Pull request by DenisRybas**
Summary:
- Model version can be deleted by the owner Vendor only before it is certified;
- When a Model is deleted, the corresponding model versions deleted as well;
- Model can be deleted by the owner Vendor only before at least one of its Model Versions is certified.

Link to ticket: https://github.com/zigbee-alliance/distributed-compliance-ledger/issues/421

### CODE REVIEW

File: docs/transactions.md [hunk 1]
```diff
@@ -256,9 +256,11 @@ All non-edited fields remain the same.
 Deletes an existing Model identified by a unique combination of `vid` (vendor ID) and `pid` (product ID)
 by the vendor account.
 
+If one of Model Versions associated with the Model is certified than Model can not be deleted. If Model has no associated certified Model Versions, Model Versions will be removed with associated Model.
```
Comment (ashcherbakov) [inline docs/transactions.md:259]:
```suggestion
If one of Model Versions associated with the Model is certified then Model can not be deleted. When Model is deleted, all associated Model Versions will be deleted as well.
```

File: docs/transactions.md [hunk 2]
```diff
@@ -340,6 +342,25 @@ All non-edited fields remain the same.
...
+
+**Status: Implemented**
+
+Deletes an existing Model Version identified by a unique combination of `vid` (vendor ID), `pid` (product ID) and `softwareVersion`
+by the vendor account.
+
+Model Version can be deleted only before it is certified.
+
+- Parameters:
+  - vid: `int32` -  model version vendor ID (positive non-zero)
```
Comment (ashcherbakov) [inline docs/transactions.md:355]: It's uint16, not int32.

File: x/model/types/errors.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ var (
 	ErrMaxSVLessThanMinSV           = sdkerrors.Register(ModuleName, 521, "max software version less than min software version")
 	ErrLsfRevisionIsNotValid        = sdkerrors.Register(ModuleName, 522, "LsfRevision should monotonically increase by 1")
 	ErrLsfRevisionIsNotAllowed      = sdkerrors.Register(ModuleName, 523, "LsfRevision is not allowed if LsfURL is not present")
+	ErrModelVersionCertified        = sdkerrors.Register(ModuleName, 524, "model version certified")
```
Comment (ashcherbakov) [inline x/model/types/errors.go:27]:
```suggestion
	ErrModelVersionDeletionCertified        = sdkerrors.Register(ModuleName, 524, "model version certified and can not be deleted")
```

File: app/app.go [hunk 4]
```diff
@@ -34,9 +34,9 @@ import (
 	tmos "github.com/tendermint/tendermint/libs/os"
 	dbm "github.com/tendermint/tm-db"
 	"github.com/zigbee-alliance/distributed-compliance-ledger/docs"
+	dclcompltypes "github.com/zigbee-alliance/distributed-compliance-ledger/types/compliance"
```
Comment (ashcherbakov) [inline app/app.go:37]: Do we really need to move ALL compliance types to `distributed-compliance-ledger/types/compliance`? It seems it's enough to move `compliance_info` only (as this is the type causing loop dependencies).

File: x/model/keeper/msg_server_model_version.go [hunk 5]
```diff
@@ -135,3 +135,55 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
...
+		return nil, types.NewErrModelVersionCertified(msg.Vid, msg.Pid, modelVersion.SoftwareVersion)
+	}
+
+	// store updated model version
+	k.RemoveModelVersion(ctx, msg.Vid, msg.Pid, msg.SoftwareVersion)
+
+	return &types.MsgDeleteModelVersionResponse{}, nil
+}
+
+func (k msgServer) IsModelCertified(ctx sdk.Context, vid int32, pid int32, softwareVersion uint32) bool {
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:179]:
```suggestion
func (k msgServer) IsModelVersionCertified(ctx sdk.Context, vid int32, pid int32, softwareVersion uint32) bool {
```

File: x/model/keeper/msg_server_model_version.go [hunk 6]
```diff
@@ -135,3 +135,55 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
...
+	certificationTypes := []string{"zigbee", "matter"}
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:180]: Please use existing `CertificationTypesList` constant, not these hard-coded values (actually this is not all supported certification types, there are 4 values in CertificationTypesList).

File: x/model/keeper/msg_server_model_version.go [hunk 7]
```diff
@@ -135,3 +135,55 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
...
+		msg.Vid,
+		msg.Pid,
+		msg.SoftwareVersion,
+	)
+
+	if !isFound {
+		return nil, types.NewErrModelVersionDoesNotExist(msg.Vid, msg.Pid, msg.SoftwareVersion)
+	}
+
+	if msg.Creator != modelVersion.Creator {
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:163]: This check is wrong. `checkModelRights` is enough. A different Vendor account with the same VID as the account who created the model (model version), can update/delete this model (model version).
Please add a corresponding test to make sure it's true.

File: x/model/types/errors.go [hunk 8]
```diff
@@ -124,3 +125,9 @@ func NewErrLsfRevisionIsNotValid(previousLsfVersion interface{},
 		"LsfRevision %v should be greater then existing lsfRevision %v by 1",
 		currentLsfVersion, previousLsfVersion)
 }
+
+func NewErrModelVersionCertified(vid interface{}, pid interface{}, softwareVersion interface{}) error {
+	return sdkerrors.Wrapf(ErrModelVersionCertified,
+		"Model version associated with vid=%v, pid=%v and softwareVersion=%v certified and its model versions can't be deleted",
```
Comment (ashcherbakov) [inline x/model/types/errors.go:131]:
```suggestion
		"Model version associated with vid=%v, pid=%v and softwareVersion=%v is certified and can't be deleted",
```

**Review by ashcherbakov — CHANGES_REQUESTED** [review]

File: x/model/keeper/msg_server_model.go [hunk 9]
```diff
@@ -181,10 +182,40 @@ func (k msgServer) DeleteModel(goCtx context.Context, msg *types.MsgDeleteModel)
...
 		return nil, types.NewErrModelDoesNotExist(msg.Vid, msg.Pid)
 	}
 
+	modelVersions, found := k.GetModelVersions(ctx, msg.Vid, msg.Pid)
+
+	if found {
+		// check if no model version has certification record
+		for _, softwareVersion := range modelVersions.SoftwareVersions {
+			if k.IsModelCertified(ctx, msg.Vid, msg.Pid, softwareVersion) {
+				return nil, types.NewErrModelVersionCertified(msg.Vid, msg.Pid, softwareVersion)
```
[the remark this answers is not in the document]
Comment (DenisRybas) [reply x/model/keeper/msg_server_model.go]: Fixed

File: x/model/keeper/msg_server_model_version.go (hunk 7 above)
Comment (DenisRybas) [reply x/model/keeper/msg_server_model_version.go]: Fixed, added test

File: x/model/keeper/msg_server_model_version.go (hunk 6 above)
Comment (DenisRybas) [reply x/model/keeper/msg_server_model_version.go]: Fixed

File: app/app.go (hunk 4 above)
Comment (DenisRybas) [reply app/app.go]: You are right. Moved types back

File: docs/transactions.md [hunk 10]
```diff
@@ -256,9 +256,11 @@ All non-edited fields remain the same.
 Deletes an existing Model identified by a unique combination of `vid` (vendor ID) and `pid` (product ID)
 by the vendor account.
 
+If one of Model Versions associated with the Model is certified than Model can not be deleted. If Model has no associated certified Model Versions, Model Versions will be removed with associated Model.
+
 - Parameters:
-  - vid: `uint16` -  model vendor ID (positive non-zero)
-  - pid: `uint16` -  model product ID (positive non-zero)
+  - vid: `int32` -  model vendor ID (positive non-zero)
```
[the remark this answers is not in the document]
Comment (DenisRybas) [reply docs/transactions.md]: Fixed it

File: docs/transactions.md (hunk 1 above)
Comment (DenisRybas) [reply docs/transactions.md]: Committed suggestion

File: x/model/types/errors.go (hunk 3 above)
Comment (DenisRybas) [reply x/model/types/errors.go]: Committed suggestion

File: x/model/keeper/msg_server_model_version.go (hunk 5 above)
Comment (DenisRybas) [reply x/model/keeper/msg_server_model_version.go]: Committed suggestion

File: x/model/keeper/msg_server_model_version.go [hunk 11]
```diff
@@ -135,3 +136,50 @@ func (k msgServer) UpdateModelVersion(goCtx context.Context, msg *types.MsgUpdat
...
+		msg.SoftwareVersion,
+	)
+
+	if !isFound {
+		return nil, types.NewErrModelVersionDoesNotExist(msg.Vid, msg.Pid, msg.SoftwareVersion)
+	}
+
+	isCertified := k.IsModelVersionCertified(ctx, msg.Vid, msg.Pid, msg.SoftwareVersion)
+	if isCertified {
+		return nil, types.NewErrModelDeletionCertified(msg.Vid, msg.Pid, modelVersion.SoftwareVersion)
```
Comment (ashcherbakov) [inline x/model/keeper/msg_server_model_version.go:166]:
```suggestion
		return nil, types.NewErrModelVersionDeletionCertified(msg.Vid, msg.Pid, modelVersion.SoftwareVersion)
```

File: x/model/types/errors.go [hunk 12]
```diff
@@ -124,3 +125,9 @@ func NewErrLsfRevisionIsNotValid(previousLsfVersion interface{},
 		"LsfRevision %v should be greater then existing lsfRevision %v by 1",
 		currentLsfVersion, previousLsfVersion)
 }
+
+func NewErrModelDeletionCertified(vid interface{}, pid interface{}, softwareVersion interface{}) error {
```
Comment (ashcherbakov) [inline x/model/types/errors.go:129]: We also need a second method NewErrModelVersionDeletionCertified with error message
`Model version associated with vid=%v, pid=%v and softwareVersion=%v has a compliance record and can't be deleted`

File: x/model/types/errors.go [hunk 13]
```diff
@@ -124,3 +125,9 @@ func NewErrLsfRevisionIsNotValid(previousLsfVersion interface{},
...
+	return sdkerrors.Wrapf(ErrModelVersionDeletionCertified,
+		"Model version associated with vid=%v, pid=%v and softwareVersion=%v is certified and the corresponding model can't be deleted",
```
Comment (ashcherbakov) [inline x/model/types/errors.go:131]:
```suggestion
		"Model version associated with vid=%v, pid=%v and softwareVersion=%v has a compliance record and the corresponding model can't be deleted",
```

File: x/model/types/errors.go [hunk 14]
```diff
@@ -11,19 +11,20 @@ var (
...
+	ErrOtaMissingInformation         = sdkerrors.Register(ModuleName, 515, "OTA missing information")
+	ErrReleaseNotesURLInvalid        = sdkerrors.Register(ModuleName, 516, "release notes URL invalid")
+	ErrModelVersionDoesNotExist      = sdkerrors.Register(ModuleName, 517, "model version does not exist")
+	ErrNoModelVersionsExist          = sdkerrors.Register(ModuleName, 518, "no model versions exist")
+	ErrModelVersionAlreadyExists     = sdkerrors.Register(ModuleName, 519, "model version already exists")
+	ErrOtaURLCannotBeSet             = sdkerrors.Register(ModuleName, 520, "OTA URL cannot be set")
+	ErrMaxSVLessThanMinSV            = sdkerrors.Register(ModuleName, 521, "max software version less than min software version")
+	ErrLsfRevisionIsNotValid         = sdkerrors.Register(ModuleName, 522, "LsfRevision should monotonically increase by 1")
+	ErrLsfRevisionIsNotAllowed       = sdkerrors.Register(ModuleName, 523, "LsfRevision is not allowed if LsfURL is not present")
+	ErrModelVersionDeletionCertified = sdkerrors.Register(ModuleName, 524, "model version certified and can not be deleted")
```
Comment (ashcherbakov) [inline x/model/types/errors.go:27]:
```suggestion
	ErrModelVersionDeletionCertified = sdkerrors.Register(ModuleName, 524, "model version has a compliance record and can not be deleted")
```

**Comment by ashcherbakov** [comment]
Also, it seems VUE files are not updated with the new Delete Model Version command. I believe we need to update them to let this action appear on the Web UI.

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**